test/unit/test_util.py

```diff
@@ -0,0 +1,229 @@
+    """
+
+    def get_num_pngs_in_dir(dir_path):
+        return len([x for x in os.listdir(dir_path) if x.endswith(".png")])
+
+    pair_dir = "logs/test_util_save_array"
+
+    # test 3D tf tensor
+    name = "3d_tf"
+    out_dir = os.path.join(pair_dir, name)
+    arr = tf.random.uniform(shape=(2, 3, 4))
+    save_array(pair_dir=pair_dir, arr=arr, name=name, gray=True)
+    assert get_num_pngs_in_dir(out_dir) == 4
+    shutil.rmtree(out_dir)
+
+    # test 4D tf tensor
+    name = "4d_tf"
+    out_dir = os.path.join(pair_dir, name)
+    arr = tf.random.uniform(shape=(2, 3, 4, 3))
+    save_array(pair_dir=pair_dir, arr=arr, name=name, gray=True)
+    assert get_num_pngs_in_dir(out_dir) == 4
+    shutil.rmtree(out_dir)
+
+    # test 3D np tensor
+    name = "3d_np"
+    out_dir = os.path.join(pair_dir, name)
+    arr = np.random.rand(2, 3, 4)
+    save_array(pair_dir=pair_dir, arr=arr, name=name, gray=True)
+    assert get_num_pngs_in_dir(out_dir) == 4
+    shutil.rmtree(out_dir)
+
+    # test 4D np tensor
+    name = "4d_np"
+    out_dir = os.path.join(pair_dir, name)
+    arr = np.random.rand(2, 3, 4, 3)
+    save_array(pair_dir=pair_dir, arr=arr, name=name, gray=True)
+    assert get_num_pngs_in_dir(out_dir) == 4
+    shutil.rmtree(out_dir)
+
+    # test 5D np tensor
+    name = "5d_np"
+    arr = np.random.rand(2, 3, 4, 1, 3)
+    with pytest.raises(ValueError) as exec_info:
+        save_array(pair_dir=pair_dir, arr=arr, name=name, gray=True)
+    msg = " ".join(exec_info.value.args[0].split())
+    assert "arr must be 3d or 4d numpy array or tf tensor" in msg
+
+    # test 4D np tensor with wrong shape
+    name = "5d_np"
+    arr = np.random.rand(2, 3, 4, 1)
+    with pytest.raises(ValueError) as exec_info:
+        save_array(pair_dir=pair_dir, arr=arr, name=name, gray=True)
+    msg = " ".join(exec_info.value.args[0].split())
+    assert "4d arr must have 3 channels as last dimension" in msg
+
+
+def test_calculate_metrics():
+    """
+    Test calculate_metrics by checking output keys.
+    Assuming the metrics functions are correct.
+    """
+
+    batch_size = 2
+    fixed_image_shape = (4, 4, 4)  # (f_dim1, f_dim2, f_dim3)
+
+    fixed_image = tf.random.uniform(shape=(batch_size,) + fixed_image_shape)
+    fixed_label = tf.random.uniform(shape=(batch_size,) + fixed_image_shape)
+    pred_moving_image = tf.random.uniform(shape=(batch_size,) + fixed_image_shape)
+    pred_fixed_label = tf.random.uniform(shape=(batch_size,) + fixed_image_shape)
+    fixed_grid_ref = tf.random.uniform(shape=(1,) + fixed_image_shape + (3,))
+    sample_index = 0
+
+    # labeled and have pred_moving_image
+    got = calculate_metrics(
+        fixed_image=fixed_image,
+        fixed_label=fixed_label,
+        pred_moving_image=pred_moving_image,
+        pred_fixed_label=pred_fixed_label,
+        fixed_grid_ref=fixed_grid_ref,
+        sample_index=sample_index,
+    )
+    assert got["image_ssd"] is not None
```

Comment:
for the `is not None` statements is there a way to check the output instead?

Reply:
In line 147, I assumed the metric functions are correct, so checking the output numerically is not my objective here. what do you think, is it ok/reasonable?